docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,341 @@
+could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another
+account.
+
+`QueryClient`s could also be made with `ModuleKey`s, except that authentication isn't required.
+
+Here's an example of a hypothetical module `foo` interacting with `x/bank`:
+```go
+package foo
+
+func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
+  bankQueryClient := bank.NewQueryClient(fooMsgServer.moduleKey)
+  balance, err := bankQueryClient.Balance(&bank.QueryBalanceRequest{Address: fooMsgServer.moduleKey.Address(), Denom: "foo"})
+  
+  ...
+
+  bankMsgClient := bank.NewMsgClient(fooMsgServer.moduleKey)
+  res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: fooMsgServer.moduleKey.Address(), ...})
+
+  ...
+}
+```
+
+### `ModuleKey`s and `ModuleID`s
+
+A `ModuleKey` can be thought of as a "private key" for a module account and a `ModuleID` can be thought of as the
+corresponding "public key". From [the ADR 028 draft](https://github.com/cosmos/cosmos-sdk/pull/7086), modules can have both a root module account and any number of sub-accounts
+or derived accounts that can be used for different pools (ex. staking pools) or managed accounts (ex. group
+accounts). We can also think of module sub-accounts as similar to derived keys - there is a root key and then some
+derivation path. `ModuleID` is a simple struct which contains the module name and optional "derivation" path,
+and forms its address based on the `AddressHash` method from [the ADR 028 draft](https://github.com/cosmos/cosmos-sdk/pull/7086):
+
+```go
+type ModuleID struct {
+  ModuleName string
+  Path []byte
+}
+
+func (key ModuleID) Address() []byte {
+  return AddressHash(key.ModuleName, key.Path)
+}
+```
+
+In addition to being able to generate a `ModuleID` and address, a `ModuleKey` contains a special function closure called
+the `Invoker` which is the key to safe inter-module access. This function closure corresponds to the `Invoke` method in
+the `grpc.ClientConn` interface and under the hood is able to route messages to the appropriate `Msg` and `Query` handlers
+performing appropriate security checks on `Msg`s. This allows for even safer inter-module access than keeper's whose
+private member variables could be manipulated through reflection. Golang does not support reflection on a function
+closure's captured variables and direct manipulation of memory would be needed for a truly malicious module to bypass
+the `ModuleKey` security.
+
+The two `ModuleKey` types are `RootModuleKey` and `DerivedModuleKey`:
+
+```go
+func Invoker(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error
+
+type CallInfo {
+  Method string
+  Caller ModuleID
+}
+
+type RootModuleKey struct {
+  moduleName string
+  invoker Invoker
+}
```

Comment:
Maybe we can make a `ModuleKey` an interface as well:
```
type ModuleKey {
    Name() string
    MkInvoke(callInfo CallInfo) grpc.Invoke  /// or InvokeFn or other appropriate type --> see comments above
    Derive(path []byte) ModuleKey
}
```

Reply:
I think we also need a cryptographic key. Using only name for authentication is not secure - other modules could use the same name.

Reply:
That's not actually true if you study the underlying design. The name doesn't do any authentication, the function closure simply restricts access.

Reply:
@robert-zaremba `ModuleKey` interface would rather look like:
```go
type ModuleKey interface {
	grpc.ClientConnInterface

	ModuleID() ModuleID
	Address() []byte
}
``` 
`grpc.ClientConnInterface` already defines an `Invoke` method.
`Derive()` method only makes sense for root `ModuleKey`s.

Reply:
Yes, looks better. I made a draft based on what's in this ADR.